client/lib/protect-form/index.jsx

```diff
@@ -0,0 +1,79 @@
+/**
+ * External dependencies
+ */
+import React, { Component } from 'react';
+import debugModule from 'debug';
+import page from 'page';
+import i18n from 'i18n-calypso';
+import { includes, without } from 'lodash';
+
+/**
+ * Module variables
+ */
+const debug = debugModule( 'calypso:protect-form' );
+const confirmText = i18n.translate( 'You have unsaved changes. Are you sure you want to leave this page?' );
+const beforeUnloadText = i18n.translate( 'You have unsaved changes.' );
+let formsChanged = [];
+
+export const protectForm = WrappedComponent => {
+	return class ProtectedFormComponent extends Component {
+		componentDidMount() {
+			window.addEventListener( 'beforeunload', this.warnIfChanged );
+		}
+
+		componentWillUnmount() {
+			window.removeEventListener( 'beforeunload', this.warnIfChanged );
+			formsChanged = without( formsChanged, this );
+		}
+
+		warnIfChanged = event => {
+			if ( ! formsChanged.length ) {
+				return;
+			}
+			debug( 'unsaved form changes detected' );
+			( event || window.event ).returnValue = beforeUnloadText;
+			return beforeUnloadText;
+		};
+
+		markChanged = () => {
+			if ( ! includes( formsChanged, this ) ) {
+				formsChanged.push( this );
+			}
+		};
+
+		markSaved = () => {
+			const index = formsChanged.indexOf( this );
```

Comment:
Since you're already pulling it in, wouldn't `without` simplify this?

```js
formsChanged = without( formsChanged, this );
```

(Similar as in `componentWillUnmount`)